### ISSUE TITLE
Prososy performance (mod_websocket and mod_bosh?)

### ISSUE THREAD

**Issue by JohnXLivingston** [JohnXLivingston/peertube-plugin-livechat#178]
When hundred of users are connected, we can see a huge amount of CPU usage form the Prosody process. We did not expect that. People working on Prosody told me they have similar rooms running with very few CPU.

I think it is the Websocket and BOSH layer that are very CPU intensive. I have to investigate the mod_websocket and mod_bosh code, and check if they are possible optimizations. (it seems these modules are doing a lot of string operations, that are not always needed).

If these modules are indeed responsible, but not easily optimizable, we should try move the Websocket and BOSH layer to an external process (something like https://github.com/moparisthebest/xmpp-proxy).

- [x] add monitoring of CPU load on the testing platform
- [x] Identify what is using CPU
- [ ] Optimize responsible code, or externalize it to another process
- [x] Stress testing and report

Reference: NlNet-task-14

**Comment by JohnXLivingston** [comment]
Seems Websocket is not the CPU upsage cause. See [Test 30-websocket-vs-c2s / Run 01](https://github.com/JohnXLivingston/livechat-perf-test/tree/main/tests/30-websocket-vs-c2s#01) for more information.

**Comment by JohnXLivingston** [comment]
It could more likely be the vCards/Avatars that are responsible of huge CPU loads.
See [Test 31-browser-and-c2s](https://github.com/JohnXLivingston/livechat-perf-test/tree/main/tests/31-browser-and-c2s#31-browser-and-c2s).
More investigation needed to confirm.

**Comment by JohnXLivingston** [comment]
Stress Testing using Puppeteer/Chromium was not possible, because it requires too much RAM and CPU on the tests hosts, and it is too slow.
The bots i had in my [testing tools](https://github.com/JohnXLivingston/livechat-perf-test) were not relevant, as there were not doing all the stuff ConverseJS does when joining the chat.
So, i added the ability to my bots to emulate some behavior:

* retrieve [disco](https://xmpp.org/extensions/xep-0030.html) info
* retrieve the user's roster
* retrieve other user's vCards (for avatar management)
* enabling [message carbon](https://xmpp.org/extensions/xep-0280.html)
* retrieve room last messages (using [MAM](https://xmpp.org/extensions/xep-0313.html))

Then i compared the load on the server, when using puppeteer/chromium, and when using bots emulating ConverseJS.

As shown in the [test 05](https://github.com/JohnXLivingston/livechat-perf-test/tree/main/tests/05-test-bot-emulation), results are very similar. So i can use bots for further testing.

**Comment by JohnXLivingston** [comment]
After discussion with some Prosody maintainers, the Lua Garbage Collector can also have some impact on the performance.

I did some tests:
* https://github.com/JohnXLivingston/peertube-plugin-livechat/issues/298
* https://github.com/JohnXLivingston/livechat-perf-test/tree/main/tests/33-prosody-gc

For now, i can't see any major benefits to change the GC parameters.
I opened #300 to remember that if i upgrade to Prosody 0.12.4+Lua5.4, i also need some tweaking.
